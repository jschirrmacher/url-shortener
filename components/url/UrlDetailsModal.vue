<script setup lang="ts">
import { ref, watch, nextTick } from "vue"
import type { UrlRecord, UpdateUrlResponse } from "~/types/index"

interface Props {
  shortCode: string
  shortUrl?: string
  originalUrl?: string
  title?: string
  isOpen: boolean
}

interface Emits {
  close: []
  updated: [url: UrlRecord]
}

const props = defineProps<Props>()
const emit = defineEmits<Emits>()

const dialog = ref<HTMLDialogElement>()
const isDialogMounted = ref(false)

// QR Code state
const qrCodeUrl = ref("")

// Update state
const updateLoading = ref(false)
const updateError = ref("")
const updateSuccess = ref(false)

// Generate QR Code
async function generateQrCode() {
  if (!props.shortCode) return
  
  try {
    // Use the correct API route and get the image directly
    qrCodeUrl.value = `/api/qr/${props.shortCode}?format=png&size=200`
  } catch (err) {
    console.error("Failed to generate QR code:", err)
  }
}

// Download QR Code
async function downloadQrCode(format: 'png' | 'svg') {
  if (!props.shortCode) return
  
  try {
    const response = await fetch(`/api/qr/${props.shortCode}?format=${format}&download=true`)
    const blob = await response.blob()
    const url = URL.createObjectURL(blob)
    const a = document.createElement('a')
    a.href = url
    a.download = `qr-${props.shortCode}.${format}`
    a.click()
    URL.revokeObjectURL(url)
  } catch (err) {
    console.error("Failed to download QR code:", err)
  }
}

// Update URL
async function updateUrl(data: { shortCode: string; originalUrl: string; title: string }) {
  if (!props.shortCode || !data.originalUrl.trim()) return

  try {
    updateLoading.value = true
    updateError.value = ""
    updateSuccess.value = false

    const body: { originalUrl: string; title?: string; newShortCode?: string } = { 
      originalUrl: data.originalUrl.trim() 
    }
    
    if (data.title !== undefined) {
      body.title = data.title.trim()
    }
    
    if (data.shortCode && data.shortCode !== props.shortCode) {
      body.newShortCode = data.shortCode
    }

    const { data: response, error } = await useFetch<UpdateUrlResponse>(`/api/urls/${props.shortCode}`, {
      method: "PUT",
      body,
    })

    if (error.value) {
      throw error.value
    }

    if (response.value?.success && response.value.url) {
      updateSuccess.value = true
      
      const updatedUrl: UrlRecord = {
        shortCode: response.value.url.shortCode,
        originalUrl: response.value.url.originalUrl,
        title: response.value.url.title || "",
        createdAt: new Date().toISOString(),
        createdBy: "current-user",
      }
      
      emit("updated", updatedUrl)
      
      setTimeout(() => {
        closeModal()
      }, 1500)
    }
  } catch (err) {
    const errorObj = err as Error & { data?: { detail?: string } }
    updateError.value = errorObj?.data?.detail ?? errorObj?.message ?? "Fehler beim Aktualisieren der URL"
  } finally {
    updateLoading.value = false
  }
}

function resetUpdate() {
  updateLoading.value = false
  updateError.value = ""
  updateSuccess.value = false
}

function openModal() {
  if (dialog.value && !dialog.value.open) {
    dialog.value.showModal()
    generateQrCode()
    nextTick(() => {
      isDialogMounted.value = true
    })
  }
}

function closeModal() {
  if (dialog.value) {
    dialog.value.close()
    resetUpdate()
    qrCodeUrl.value = ""
    isDialogMounted.value = false
    emit("close")
  }
}

function handleBackdropClick(event: MouseEvent) {
  if (event.target === dialog.value) {
    closeModal()
  }
}

function handleKeydown(event: KeyboardEvent) {
  if (event.key === "Escape") {
    closeModal()
  }
}

// Watch for isOpen prop changes
watch(
  () => props.isOpen,
  (isOpen) => {
    if (isOpen) {
      openModal()
    } else {
      closeModal()
    }
  },
)
</script>

<template>
  <dialog
    ref="dialog"
    class="modal-dialog"
    @click="handleBackdropClick"
    @keydown="handleKeydown"
  >
    <div class="modal-content">
      <ModalHeader 
        :title="`URL Details: ${shortCode}`"
        @close="closeModal"
      />

      <div class="modal-body">
        <QrCodeSection
          :short-code="shortCode"
          :qr-code-url="qrCodeUrl"
          @download="downloadQrCode"
        />

        <AlertMessage
          v-if="updateError"
          type="error"
          title="Fehler beim Aktualisieren"
          :message="updateError"
        />

        <AlertMessage
          v-if="updateSuccess"
          type="success"
          title="Erfolgreich aktualisiert"
          message="Die URL wurde erfolgreich geÃ¤ndert."
        />

        <UrlForm
          :short-url="shortUrl || ''"
          :short-code="shortCode"
          :original-url="originalUrl || ''"
          :title="title"
          @changed="updateUrl"
          @cancel="closeModal"
        />
      </div>
    </div>
  </dialog>
</template>

<style scoped>
.modal-dialog {
  border: none;
  border-radius: 0.5rem;
  box-shadow: 0 20px 25px -5px rgb(0 0 0 / 0.1), 0 8px 10px -6px rgb(0 0 0 / 0.1);
  padding: 0;
  max-width: 32rem;
  width: 100%;
  max-height: 90vh;
}

.modal-dialog::backdrop {
  background-color: rgba(0, 0, 0, 0.5);
}

.modal-content {
  background-color: #ffffff;
  border-radius: 0.5rem;
  max-height: 90vh;
  overflow: hidden;
  display: flex;
  flex-direction: column;
}

.modal-body {
  padding: 1.5rem;
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
  overflow-y: auto;
}
</style>
