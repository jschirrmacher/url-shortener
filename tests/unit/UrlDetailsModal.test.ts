/**
 * @vitest-environment happy-dom
 */
import { describe, it, expect, vi, beforeEach } from "vitest"
import { mount } from "@vue/test-utils"
import UrlDetailsModal from "~/components/url/UrlDetailsModal.vue"

// Mock global functions
const mockFetch = vi.fn()
global.$fetch = mockFetch as any

// Mock useRuntimeConfig
vi.mock('#app', () => ({
  useRuntimeConfig: () => ({
    public: {
      baseUrl: 'http://localhost:3000'
    }
  })
}))

describe("UrlDetailsModal Component", () => {
  const mockUrl = {
    shortCode: "abc123",
    originalUrl: "https://example.com",
    title: "Example Site",
    createdAt: "2024-01-01T00:00:00.000Z",
    createdBy: "testuser",
    totalClicks: 42
  }

  beforeEach(() => {
    vi.clearAllMocks()
  })

  it.skip("should render URL details correctly", () => {
    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    expect(wrapper.text()).toContain("abc123")
    expect(wrapper.text()).toContain("https://example.com")
    expect(wrapper.text()).toContain("Example Site")
  })

  it.skip("should show form fields when editing", async () => {
    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    // Find and click edit button
    const editButton = wrapper.find('[data-testid="edit-button"]')
    if (editButton.exists()) {
      await editButton.trigger('click')
    }

    // Should show input fields
    expect(wrapper.find('input[type="url"]').exists()).toBe(true)
    expect(wrapper.find('input[placeholder*="Short Code"]').exists()).toBe(true)
  })

  it("should validate shortCode format", async () => {
    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    // Enter edit mode
    const editButton = wrapper.find('[data-testid="edit-button"]')
    if (editButton.exists()) {
      await editButton.trigger('click')
    }

    // Try invalid shortCode
    const shortCodeInput = wrapper.find('input[placeholder*="Short Code"]')
    if (shortCodeInput.exists()) {
      await shortCodeInput.setValue('invalid code') // space not allowed
      
      // Should show validation error
      expect(wrapper.text()).toContain('Nur Buchstaben, Zahlen, Bindestriche und Unterstriche erlaubt')
    }
  })

  it.skip("should call API when updating URL", async () => {
    const mockResponse = {
      success: true,
      message: "URL erfolgreich aktualisiert",
      url: {
        ...mockUrl,
        originalUrl: "https://newexample.com"
      }
    }
    mockFetch.mockResolvedValueOnce(mockResponse)

    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    // Enter edit mode
    const editButton = wrapper.find('[data-testid="edit-button"]')
    if (editButton.exists()) {
      await editButton.trigger('click')
    }

    // Change URL
    const urlInput = wrapper.find('input[type="url"]')
    if (urlInput.exists()) {
      await urlInput.setValue('https://newexample.com')
    }

    // Submit form
    const saveButton = wrapper.find('[data-testid="save-button"]')
    if (saveButton.exists()) {
      await saveButton.trigger('click')
    }

    // Should call API
    expect(mockFetch).toHaveBeenCalledWith(`/api/urls/${mockUrl.shortCode}`, {
      method: 'PUT',
      body: {
        originalUrl: 'https://newexample.com'
      }
    })
  })

  it.skip("should include newShortCode when shortCode changes", async () => {
    const mockResponse = {
      success: true,
      message: "URL erfolgreich aktualisiert",
      url: {
        ...mockUrl,
        shortCode: "new123"
      }
    }
    mockFetch.mockResolvedValueOnce(mockResponse)

    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    // Enter edit mode
    const editButton = wrapper.find('[data-testid="edit-button"]')
    if (editButton.exists()) {
      await editButton.trigger('click')
    }

    // Change shortCode
    const shortCodeInput = wrapper.find('input[placeholder*="Short Code"]')
    if (shortCodeInput.exists()) {
      await shortCodeInput.setValue('new123')
    }

    // Submit form
    const saveButton = wrapper.find('[data-testid="save-button"]')
    if (saveButton.exists()) {
      await saveButton.trigger('click')
    }

    // Should call API with newShortCode
    expect(mockFetch).toHaveBeenCalledWith(`/api/urls/${mockUrl.shortCode}`, {
      method: 'PUT',
      body: {
        originalUrl: mockUrl.originalUrl,
        newShortCode: 'new123'
      }
    })
  })

  it.skip("should show error message on API failure", async () => {
    const errorMessage = "Short Code bereits vergeben"
    mockFetch.mockRejectedValueOnce({
      data: { message: errorMessage }
    })

    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    // Enter edit mode and try to save
    const editButton = wrapper.find('[data-testid="edit-button"]')
    if (editButton.exists()) {
      await editButton.trigger('click')
    }

    const saveButton = wrapper.find('[data-testid="save-button"]')
    if (saveButton.exists()) {
      await saveButton.trigger('click')
    }

    // Wait for error to appear
    await wrapper.vm.$nextTick()

    // Should show error message
    expect(wrapper.text()).toContain(errorMessage)
  })

  it.skip("should emit close event when modal is closed", async () => {
    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    // Find and click close button
    const closeButton = wrapper.find('[data-testid="close-button"]')
    if (closeButton.exists()) {
      await closeButton.trigger('click')
    }

    // Should emit close event
    expect(wrapper.emitted('close')).toBeTruthy()
  })

  it.skip("should emit updated event when URL is successfully updated", async () => {
    const mockResponse = {
      success: true,
      message: "URL erfolgreich aktualisiert",
      url: {
        ...mockUrl,
        originalUrl: "https://newexample.com"
      }
    }
    mockFetch.mockResolvedValueOnce(mockResponse)

    const wrapper = mount(UrlDetailsModal, {
      props: {
        url: mockUrl,
        isOpen: true
      }
    })

    // Simulate successful update
    const editButton = wrapper.find('[data-testid="edit-button"]')
    if (editButton.exists()) {
      await editButton.trigger('click')
    }

    const saveButton = wrapper.find('[data-testid="save-button"]')
    if (saveButton.exists()) {
      await saveButton.trigger('click')
    }

    // Wait for API call to complete
    await wrapper.vm.$nextTick()

    // Should emit updated event with new URL data
    expect(wrapper.emitted('updated')).toBeTruthy()
    expect(wrapper.emitted('updated')?.[0]).toEqual([mockResponse.url])
  })
})
