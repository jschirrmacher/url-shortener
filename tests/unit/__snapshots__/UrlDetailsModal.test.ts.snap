// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UrlDetailsModal Component > should match snapshot 1`] = `
"<dialog data-v-dd4104c0="" class="modal-dialog shadow-xl">
  <div data-v-dd4104c0="" class="modal-content container-primary">
    <modal-header-stub data-v-dd4104c0="" title="URL Details: abc123"></modal-header-stub>
    <div data-v-dd4104c0="" class="modal-body">
      <qr-code-section-stub data-v-dd4104c0="" short-code="abc123" qr-code-url=""></qr-code-section-stub>
      <div data-v-dd4104c0="" class="form-section">
        <!--v-if-->
        <!--v-if-->
        <url-form-stub data-v-dd4104c0="" short-url="http://localhost:3000/abc123" short-code="abc123" original-url="https://example.com" title="Example Website" owner="testuser"></url-form-stub>
      </div>
    </div>
  </div>
</dialog>"
`;
